File: packages/less/src/less/tree/node.go
package tree

import (
	"fmt"
	"math"
	"strings"
)

// Node represents a base node in the Less AST
type Node struct {
	Parent          *Node
	VisibilityBlocks *int
	NodeVisible     *bool
	RootNode        *Node
	Parsed          interface{}
	Value           interface{}
	Index           int
	fileInfo        map[string]interface{}
}

// NewNode creates a new Node instance
func NewNode() *Node {
	return &Node{
		Parent:          nil,
		VisibilityBlocks: nil,
		NodeVisible:     nil,
		RootNode:        nil,
		Parsed:          nil,
		Value:           nil,
		Index:           0,
		fileInfo:        make(map[string]interface{}),
	}
}

// SetParent sets the parent for one or more nodes
func (n *Node) SetParent(nodes interface{}, parent *Node) {
	switch v := nodes.(type) {
	case []*Node:
		for _, node := range v {
			if node != nil {
				node.Parent = parent
			}
		}
	case *Node:
		if v != nil {
			v.Parent = parent
		}
	case nil:
		// Handle nil case gracefully
	default:
		// Handle non-Node types gracefully
	}
}

// GetIndex returns the node's index
func (n *Node) GetIndex() int {
	if n.Index != 0 {
		return n.Index
	}
	if n.Parent != nil {
		return n.Parent.GetIndex()
	}
	return 0
}

// FileInfo returns the node's file information
func (n *Node) FileInfo() map[string]interface{} {
	if len(n.fileInfo) > 0 {
		return n.fileInfo
	}
	if n.Parent != nil {
		return n.Parent.FileInfo()
	}
	return make(map[string]interface{})
}

// SetFileInfo sets the node's file information
func (n *Node) SetFileInfo(info map[string]interface{}) {
	n.fileInfo = info
}

// IsRulesetLike returns false for base Node
func (n *Node) IsRulesetLike() bool {
	return false
}

// ToCSS generates CSS string representation
func (n *Node) ToCSS(context interface{}) string {
	var strs []string
	output := &CSSOutput{
		Add: func(chunk interface{}, fileInfo interface{}, index interface{}) {
			strs = append(strs, fmt.Sprintf("%v", chunk))
		},
		IsEmpty: func() bool {
			return len(strs) == 0
		},
	}
	n.GenCSS(context, output)
	return strings.Join(strs, "")
}

// CSSOutput represents the output structure for CSS generation
type CSSOutput struct {
	Add     func(interface{}, interface{}, interface{})
	IsEmpty func() bool
}

// GenCSS generates CSS representation
func (n *Node) GenCSS(context interface{}, output *CSSOutput) {
	if n.Value != nil {
		output.Add(n.Value, nil, nil)
	}
}

// Visitor interface defines the Visit method
type Visitor interface {
	Visit(interface{}) interface{}
}

// Accept visits the node with a visitor
func (n *Node) Accept(visitor interface{}) {
	if v, ok := visitor.(Visitor); ok {
		if n.Value != nil {
			n.Value = v.Visit(n.Value)
		}
	}
}

// Eval returns the node itself
func (n *Node) Eval() *Node {
	return n
}

// Operate performs basic arithmetic operations
func (n *Node) Operate(context interface{}, op string, a, b float64) float64 {
	switch op {
	case "+":
		return a + b
	case "-":
		return a - b
	case "*":
		return a * b
	case "/":
		return a / b
	default:
		return 0
	}
}

// Fround rounds numbers based on precision
func (n *Node) Fround(context interface{}, value float64) float64 {
	if context == nil {
		return value
	}
	
	// Add "epsilon" to ensure numbers like 1.000000005 are properly rounded
	epsilon := 2e-16
	if ctx, ok := context.(map[string]interface{}); ok {
		if precision, ok := ctx["numPrecision"].(int); ok {
			// Use math.Round to match JavaScript's toFixed behavior
			rounded := math.Round((value+epsilon)*math.Pow10(precision)) / math.Pow10(precision)
			return rounded
		}
	}
	return value
}

// Compareable interface defines the Compare method
type Compareable interface {
	Compare(*Node) int
}

// Compare compares two nodes
func Compare(a, b *Node) int {
	if a == nil || b == nil {
		return 0
	}

	// Handle nodes with Compare method
	if aCompare, ok := a.Value.(Compareable); ok {
		// For "symmetric results" force toCSS-based comparison
		// of Quoted or Anonymous if either value is one of those
		if bType, ok := b.Value.(string); ok {
			if bType == "Quoted" || bType == "Anonymous" {
				return strings.Compare(fmt.Sprintf("%v", a.Value), fmt.Sprintf("%v", b.Value))
			}
		}
		return aCompare.Compare(b)
	}
	if bCompare, ok := b.Value.(Compareable); ok {
		return -bCompare.Compare(a)
	}

	// Check type equality
	if fmt.Sprintf("%T", a.Value) != fmt.Sprintf("%T", b.Value) {
		// If types don't match, check if one is a string and the other is a number
		if aStr, ok := a.Value.(string); ok {
			if bNum, ok := b.Value.(float64); ok {
				return strings.Compare(aStr, fmt.Sprintf("%v", bNum))
			}
		}
		if bStr, ok := b.Value.(string); ok {
			if aNum, ok := a.Value.(float64); ok {
				return strings.Compare(fmt.Sprintf("%v", aNum), bStr)
			}
		}
		return 0 // Different types should return 0 (false) for equality comparison
	}

	// Handle array comparison
	if aArr, ok := a.Value.([]interface{}); ok {
		if bArr, ok := b.Value.([]interface{}); ok {
			if len(aArr) != len(bArr) {
				return 0 // Equivalent to JavaScript's undefined
			}
			for i := range aArr {
				if Compare(&Node{Value: aArr[i]}, &Node{Value: bArr[i]}) != 0 {
					return 0 // Equivalent to JavaScript's undefined
				}
			}
			return 0
		}
	}

	// Handle simple value comparison
	if a.Value == b.Value {
		return 0
	}

	// Handle numeric comparison
	if aNum, ok := a.Value.(float64); ok {
		if bNum, ok := b.Value.(float64); ok {
			return NumericCompare(aNum, bNum)
		}
	}

	// Handle string comparison
	if aStr, ok := a.Value.(string); ok {
		if bStr, ok := b.Value.(string); ok {
			// Special handling for Quoted and Anonymous types
			if aStr == "Quoted" || aStr == "Anonymous" || bStr == "Quoted" || bStr == "Anonymous" {
				return strings.Compare(aStr, bStr)
			}
			return strings.Compare(aStr, bStr)
		}
	}

	return 0 // Equivalent to JavaScript's undefined
}

// NumericCompare compares two numbers
func NumericCompare(a, b float64) int {
	if a < b {
		return -1
	}
	if a == b {
		return 0
	}
	if a > b {
		return 1
	}
	return 0
}

// BlocksVisibility returns true if the node blocks visibility
func (n *Node) BlocksVisibility() bool {
	if n.VisibilityBlocks == nil {
		return false
	}
	return *n.VisibilityBlocks != 0
}

// AddVisibilityBlock increments visibility blocks
func (n *Node) AddVisibilityBlock() {
	if n.VisibilityBlocks == nil {
		zero := 0
		n.VisibilityBlocks = &zero
	}
	*n.VisibilityBlocks++
}

// RemoveVisibilityBlock decrements visibility blocks
func (n *Node) RemoveVisibilityBlock() {
	if n.VisibilityBlocks != nil && *n.VisibilityBlocks > 0 {
		*n.VisibilityBlocks--
	}
}

// EnsureVisibility sets node visibility to true
func (n *Node) EnsureVisibility() {
	trueVal := true
	n.NodeVisible = &trueVal
}

// EnsureInvisibility sets node visibility to false
func (n *Node) EnsureInvisibility() {
	falseVal := false
	n.NodeVisible = &falseVal
}

// IsVisible returns the node's visibility state
func (n *Node) IsVisible() *bool {
	return n.NodeVisible
}

// VisibilityInfo returns the node's visibility information
func (n *Node) VisibilityInfo() map[string]interface{} {
	return map[string]interface{}{
		"visibilityBlocks": n.VisibilityBlocks,
		"nodeVisible":     n.NodeVisible,
	}
}

// CopyVisibilityInfo copies visibility information from another node
func (n *Node) CopyVisibilityInfo(info map[string]interface{}) {
	if info == nil {
		return
	}
	if blocks, ok := info["visibilityBlocks"].(*int); ok {
		n.VisibilityBlocks = blocks
	}
	if visible, ok := info["nodeVisible"].(*bool); ok {
		n.NodeVisible = visible
	}
} 

File: packages/less/lib/less/tree/node.js
// @ts-check
/**
 * @typedef {object} FileInfo
 * @property {string} [filename]
 * @property {string} [rootpath]
 * @property {string} [currentDirectory]
 * @property {string} [rootFilename]
 * @property {string} [entryPath]
 * @property {boolean} [reference]
 */

/**
 * @typedef {object} VisibilityInfo
 * @property {number} [visibilityBlocks]
 * @property {boolean} [nodeVisible]
 */

/**
 * @typedef {object} CSSOutput
 * @property {(chunk: string, fileInfo?: FileInfo, index?: number, mapLines?: boolean) => void} add
 * @property {() => boolean} isEmpty
 */

/**
 * @typedef {object} EvalContext
 * @property {number} [numPrecision]
 * @property {(op?: string) => boolean} [isMathOn]
 * @property {number} [math]
 * @property {Node[]} [frames]
 * @property {Array<{important?: string}>} [importantScope]
 * @property {string[]} [paths]
 * @property {boolean} [compress]
 * @property {boolean} [strictUnits]
 * @property {boolean} [sourceMap]
 * @property {boolean} [importMultiple]
 * @property {string} [urlArgs]
 * @property {boolean} [javascriptEnabled]
 * @property {object} [pluginManager]
 * @property {number} [rewriteUrls]
 * @property {boolean} [inCalc]
 * @property {boolean} [mathOn]
 * @property {boolean[]} [calcStack]
 * @property {boolean[]} [parensStack]
 * @property {Node[]} [mediaBlocks]
 * @property {Node[]} [mediaPath]
 * @property {() => void} [inParenthesis]
 * @property {() => void} [outOfParenthesis]
 * @property {() => void} [enterCalc]
 * @property {() => void} [exitCalc]
 * @property {(path: string) => boolean} [pathRequiresRewrite]
 * @property {(path: string, rootpath?: string) => string} [rewritePath]
 * @property {(path: string) => string} [normalizePath]
 * @property {number} [tabLevel]
 * @property {boolean} [lastRule]
 */

/**
 * @typedef {object} TreeVisitor
 * @property {(node: Node) => Node} visit
 * @property {(nodes: Node[], nonReplacing?: boolean) => Node[]} visitArray
 */

/**
 * The reason why Node is a class and other nodes simply do not extend
 * from Node (since we're transpiling) is due to this issue:
 *
 * @see https://github.com/less/less.js/issues/3434
 */
class Node {
    get type() { return ''; }

    constructor() {
        /** @type {Node | null} */
        this.parent = null;
        /** @type {number | undefined} */
        this.visibilityBlocks = undefined;
        /** @type {boolean | undefined} */
        this.nodeVisible = undefined;
        /** @type {Node | null} */
        this.rootNode = null;
        /** @type {Node | null} */
        this.parsed = null;

        /** @type {Node | Node[] | string | number | undefined} */
        this.value = undefined;
        /** @type {number | undefined} */
        this._index = undefined;
        /** @type {FileInfo | undefined} */
        this._fileInfo = undefined;
    }

    get currentFileInfo() {
        return this.fileInfo();
    }

    get index() {
        return this.getIndex();
    }

    /**
     * @param {Node | Node[]} nodes
     * @param {Node} parent
     */
    setParent(nodes, parent) {
        /** @param {Node} node */
        function set(node) {
            if (node && node instanceof Node) {
                node.parent = parent;
            }
        }
        if (Array.isArray(nodes)) {
            nodes.forEach(set);
        }
        else {
            set(nodes);
        }
    }

    /** @returns {number} */
    getIndex() {
        return this._index || (this.parent && this.parent.getIndex()) || 0;
    }

    /** @returns {FileInfo} */
    fileInfo() {
        return this._fileInfo || (this.parent && this.parent.fileInfo()) || {};
    }

    /** @returns {boolean} */
    isRulesetLike() { return false; }

    /**
     * @param {EvalContext} context
     * @returns {string}
     */
    toCSS(context) {
        /** @type {string[]} */
        const strs = [];
        this.genCSS(context, {
            add: function(chunk, fileInfo, index) {
                strs.push(chunk);
            },
            isEmpty: function () {
                return strs.length === 0;
            }
        });
        return strs.join('');
    }

    /**
     * @param {EvalContext} context
     * @param {CSSOutput} output
     */
    genCSS(context, output) {
        output.add(/** @type {string} */ (this.value));
    }

    /**
     * @param {TreeVisitor} visitor
     */
    accept(visitor) {
        this.value = visitor.visit(/** @type {Node} */ (this.value));
    }

    /**
     * @param {EvalContext} [context]
     * @returns {Node}
     */
    eval(context) { return this; }

    /**
     * @param {EvalContext} context
     * @param {string} op
     * @param {number} a
     * @param {number} b
     * @returns {number | undefined}
     */
    _operate(context, op, a, b) {
        switch (op) {
            case '+': return a + b;
            case '-': return a - b;
            case '*': return a * b;
            case '/': return a / b;
        }
    }

    /**
     * @param {EvalContext} context
     * @param {number} value
     * @returns {number}
     */
    fround(context, value) {
        const precision = context && context.numPrecision;
        // add "epsilon" to ensure numbers like 1.000000005 (represented as 1.000000004999...) are properly rounded:
        return (precision) ? Number((value + 2e-16).toFixed(precision)) : value;
    }

    /**
     * @param {Node & { compare?: (other: Node) => number | undefined }} a
     * @param {Node & { compare?: (other: Node) => number | undefined }} b
     * @returns {number | undefined}
     */
    static compare(a, b) {
        /* returns:
         -1: a < b
         0: a = b
         1: a > b
         and *any* other value for a != b (e.g. undefined, NaN, -2 etc.) */

        if ((a.compare) &&
            // for "symmetric results" force toCSS-based comparison
            // of Quoted or Anonymous if either value is one of those
            !(b.type === 'Quoted' || b.type === 'Anonymous')) {
            return a.compare(b);
        } else if (b.compare) {
            return -b.compare(a);
        } else if (a.type !== b.type) {
            return undefined;
        }

        let aVal = a.value;
        let bVal = b.value;
        if (!Array.isArray(aVal)) {
            return aVal === bVal ? 0 : undefined;
        }
        if (!Array.isArray(bVal)) {
            return undefined;
        }
        if (aVal.length !== bVal.length) {
            return undefined;
        }
        for (let i = 0; i < aVal.length; i++) {
            if (Node.compare(aVal[i], bVal[i]) !== 0) {
                return undefined;
            }
        }
        return 0;
    }

    /**
     * @param {number | string} a
     * @param {number | string} b
     * @returns {number | undefined}
     */
    static numericCompare(a, b) {
        return a  <  b ? -1
            : a === b ?  0
                : a  >  b ?  1 : undefined;
    }

    /** @returns {boolean} */
    blocksVisibility() {
        if (this.visibilityBlocks === undefined) {
            this.visibilityBlocks = 0;
        }
        return this.visibilityBlocks !== 0;
    }

    addVisibilityBlock() {
        if (this.visibilityBlocks === undefined) {
            this.visibilityBlocks = 0;
        }
        this.visibilityBlocks = this.visibilityBlocks + 1;
    }

    removeVisibilityBlock() {
        if (this.visibilityBlocks === undefined) {
            this.visibilityBlocks = 0;
        }
        this.visibilityBlocks = this.visibilityBlocks - 1;
    }

    ensureVisibility() {
        this.nodeVisible = true;
    }

    ensureInvisibility() {
        this.nodeVisible = false;
    }

    /** @returns {boolean | undefined} */
    isVisible() {
        return this.nodeVisible;
    }

    /** @returns {VisibilityInfo} */
    visibilityInfo() {
        return {
            visibilityBlocks: this.visibilityBlocks,
            nodeVisible: this.nodeVisible
        };
    }

    /** @param {VisibilityInfo} info */
    copyVisibilityInfo(info) {
        if (!info) {
            return;
        }
        this.visibilityBlocks = info.visibilityBlocks;
        this.nodeVisible = info.nodeVisible;
    }
}

/**
 * Set by the parser at runtime on Node.prototype.
 * @type {{ context: EvalContext, importManager: object, imports: object } | undefined}
 */
Node.prototype.parse = undefined;

export default Node;
